front/src/applications/operationalStudies/helpers/TrainSimulationLazyLoader.ts

```diff
@@ -0,0 +1,146 @@
+import {
+  osrdEditoastApi,
+  type PostPacedTrainSimulationSummaryApiResponse,
+  type PostTrainScheduleSimulationSummaryApiResponse,
+  type SimulationSummaryResult,
+} from 'common/api/osrdEditoastApi';
+import type { TimetableItemId } from 'reducers/osrdconf/types';
+import type { AppDispatch } from 'store';
+import {
+  formatEditoastTrainIdToPacedTrainId,
+  formatEditoastTrainIdToTrainScheduleId,
+  formatPacedTrainIdToEditoastTrainId,
+  formatTrainScheduleIdToEditoastTrainId,
+  isTrainSchedule,
+} from 'utils/trainId';
+
+const BATCH_SIZE = 20;
+
+type TrainSimulationLazyLoaderOptions = {
+  dispatch: AppDispatch;
+  infraId: number;
+  electricalProfileSetId?: number;
+  onProgress: (results: Map<TimetableItemId, SimulationSummaryResult>) => void;
+};
+
+/**
+ * Takes a stream of train IDs as input, incrementally invokes onProgress when
+ * train simulations become available.
+ *
+ * This helper only takes care of the API requests, it doesn't perform any
+ * post-processing of the simulation results.
+ */
+export default class TrainSimulationLazyLoader {
+  readonly options: TrainSimulationLazyLoaderOptions;
+
+  pending: TimetableItemId[] = [];
+
+  prevPromise: Promise<void> = Promise.resolve();
+
+  cancelled = false;
+
+  /**
+   * Create a new loader. Options are immutable for the lifetime of the loader.
+   */
+  constructor(options: TrainSimulationLazyLoaderOptions) {
+    this.options = options;
+  }
+
+  /**
+   * Queue train IDs for simulation.
+   */
+  simulateTimetableItems(ids: TimetableItemId[]) {
+    if (this.cancelled) {
+      throw new Error('simulateTimetableItems() called after cancel()');
+    }
+    this.pending = [...this.pending, ...ids];
+    this.prevPromise = this.prevPromise.finally(() => this.processPending());
+  }
+
+  /**
+   * Cancel all pending train simulations. The loader cannot be used after
+   * calling this method.
+   */
+  cancel() {
+    this.pending = [];
+    this.cancelled = true;
+  }
+
+  async processPending() {
+    while (this.pending.length > 0) {
+      const batch = this.pending.slice(0, BATCH_SIZE);
+      this.pending = this.pending.slice(BATCH_SIZE);
```

Comment:
creating new arrays every cycle is wasteful in JS, if we have thousands of trains I think it might make a difference. 
Here you can work in place and start taking trains from the end of the array to avoid shifting.

what do you think ?

```suggestion
  const batch = this.pending.splice(-BATCH_SIZE);
  await this.processBatch(batch);
```

Reply:
Unless there is a proof that this actually matters in a benchmark, I'd say this is premature optimization.

Reply:
This is known to use more memory, which over times will demand more garbage collection and in the end slow down the application. our application is already slow as it is, what do we lose from optimizing here ?

Reply:
I find `splice()` to be confusing and best avoided, because it does a lot of different things and figuring out what the call actually does requires pulling up the docs and carefully reading each clause. I value readability over performance, and in general I think this has very little to no impact in practice because it's just an array of a few thousand integers in the worst case.

The app is slow, but making this change will not fix that slowness. The slowness will not worsen because of this change. Fixing the slowness requires making benchmarks and interpreting CPU/memory usage graphs.